bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogic.java

```diff
@@ -0,0 +1,789 @@
+		return isActive(SearchOptions.REGEX) && fIsTargetSupportingRegEx;
+	}
+
+
+	/**
+	 * Initializes the anchor used as starting point for incremental searching.
+	 *
+	 * @since 2.0
+	 */
+	private void initIncrementalBaseLocation() {
+		if (target != null && isActive(SearchOptions.INCREMENTAL) && !isRegExSearchAvailableAndActive()) {
+			incrementalBaseLocation = target.getSelection();
+		} else {
+			incrementalBaseLocation = new Point(0, 0);
+		}
+	}
+
+	public boolean shouldInitIncrementalBaseLocation() {
+		return isActive(SearchOptions.INCREMENTAL) && !isActive(SearchOptions.REGEX);
+	}
+
+	/**
+	 * Tells the dialog to perform searches only in the scope given by the actually
+	 * selected lines.
+	 *
+	 * @param selectedLines <code>true</code> if selected lines should be used
+	 * @since 2.0
+	 */
+	@Override
+	public void useSelectedLines(boolean selectedLines) {
+		if (shouldInitIncrementalBaseLocation()) {
+			initIncrementalBaseLocation();
+		}
+
+		if (target == null || !(target instanceof IFindReplaceTargetExtension)) {
+			return;
+		}
+
+		IFindReplaceTargetExtension extensionTarget = (IFindReplaceTargetExtension) target;
+
+		if (selectedLines) {
+
+			IRegion scope;
+			if (oldScope == null) {
+				Point lineSelection = extensionTarget.getLineSelection();
+				scope = new Region(lineSelection.x, lineSelection.y);
+			} else {
+				scope = oldScope;
+				oldScope = null;
+			}
+
+			int offset = isActive(SearchOptions.FORWARD) ? scope.getOffset() : scope.getOffset() + scope.getLength();
+
+			extensionTarget.setSelection(offset, 0);
+			extensionTarget.setScope(scope);
+		} else {
+			oldScope = extensionTarget.getScope();
+			extensionTarget.setScope(null);
+		}
+	}
+
+	/**
+	 * Returns the status line manager of the active editor or <code>null</code> if
+	 * there is no such editor.
+	 *
+	 * @return the status line manager of the active editor
+	 */
+	private IEditorStatusLine getStatusLineManager() {
+		IWorkbenchWindow window = PlatformUI.getWorkbench().getActiveWorkbenchWindow();
+		if (window == null) {
+			return null;
+		}
+
+		IWorkbenchPage page = window.getActivePage();
+		if (page == null) {
+			return null;
+		}
+
+		IEditorPart editor = page.getActiveEditor();
+		if (editor == null) {
+			return null;
+		}
+
+		return editor.getAdapter(IEditorStatusLine.class);
+	}
+
+
+	/**
+	 * Replaces all occurrences of the user's findString with the replace string.
+	 * Indicate to the user the number of replacements that occur.
+	 *
+	 * <b> status: </b> {@see #getStatus() } message: String telling a user how many
+	 * replacements were performed or that the find-String could not be matched at
+	 * all
+	 *
+	 * @param findString    The string that will be replaced
+	 * @param replaceString The string that will replace the findString
+	 * @param display       the UI's Display
+	 */
+	@Override
+	public void performReplaceAll(String findString, String replaceString, Display display) {
```

Comment:
`performReplaceAll` and `performSelectAll` have a lot in common - is it possible to generalize them?

Reply:
@HeikoKlare my Idea is to create a `performOnWholeScope`-Method that takes in such a runnable. However, my concern is that the "Result handling" is too different in both methods. Additionally, the arguments these runnables take are not as similar as, adding need for extra-exception handling.

Reply:
That's a task for a separate PR. Let's keep this refactoring focused as is (extract business logic from FindReplaceDialog) and leave refactoring that is internal to the logic for a later PR.